gapfill/gapfill.py

```diff
@@ -1,22 +1,237 @@
+                                predictors_path is provided.
+                                Certain keyword predictors are used to denote
+                                specific sets of predictors:
+                                ['temporal', 'all']
+        predictors_paths (list<str>): Comma-separated list of paths to files
+                                      containing predictor names. See
+                                      predictors/metereological.txt for an
+                                      example.
+        distribution (str): Which distribution to use for prediction.
+                            Options: ['laplace', 'normal']
+        budget_date_ranges_path (str): Dictionary mapping site names to date
+                                       ranges. See
+                                           gapfill/budget_date_ranges.json
+                                       for an example.
+
+    Writes gapfilled data to
+        data/{SiteID}/gapfilled/{model}_{predictors}_{distribution}.csv
+    where the CSV has all the same columns as raw.csv, excluding any existing
+    gapfilled columns, plus columns for the
+        predicted mean (FCH4_F)
+        95% uncertainty (FCH4_uncertainty)
+        spread individual predictions (FCH4{1-N})
+    and writes budget data to
+        data/{SiteID}/gapfilled/{model}_{predictors}_{distribution}_budget.csv
     """
-    pass
\ No newline at end of file
+    data_dir = Path("data/")
+    model_dirs = parse_model_dirs(
+        model_dirs=model_dirs,
+        sites=sites,
+        models=models,
+        predictors=predictors,
+        predictors_paths=predictors_paths,
+        data_dir=data_dir
+    )
+    if budget_date_ranges_path is not None:
+        with Path(budget_date_ranges_path).open() as f:
+            budget_date_ranges = json.load(f)
+    else:
+        budget_date_ranges = {}
+
+    for model_dir in model_dirs:
+        model_dir = Path(model_dir)
+        site_data_dir = model_dir.parent.parent.parent
+
+        predictor_subset = model_dir.name
+        model = model_dir.parent.name
+        site = site_data_dir.name
+
+        gap_dir = site_data_dir / "gapfilled"
+        gap_dir.mkdir(exist_ok=True)
+
+        gap_path = site_data_dir / 'gap.csv'
+        gap_df = pd.read_csv(gap_path)
+
+        model_obj = EnsembleModel(model_dir)
+
+        scale_path = model_dir / "scale.json"
+        if not scale_path.exists():
+            raise ValueError(
+                "Must run <python main.py test> with --distribution " +
+                f"{distribution} to compute an uncertainty scale before " +
+                "gapfilling."
+            )
+        with scale_path.open() as f:
+            uncertainty_scale = json.load(f)
+
+        y_hat_individual = model_obj.predict_individual(gap_df)
+        y_hat_mean = np.mean(y_hat_individual, axis=0)
+        y_hat_spread = (
+            (y_hat_individual - y_hat_mean) * uncertainty_scale + y_hat_mean
```

Comment:
`uncertainty_scale` is the Platt scaling value, which will always be > 1 right?

Reply:
Yep it's the Platt scaling value. It can be less than 1 too if the original uncertainty was too large, but I didn't see that happen in any of our experiments.